go/terminalescaper/escaper.go

```diff
@@ -2,24 +2,114 @@ package terminalescaper
 
 import (
 	"io"
-	"strings"
+	"unicode/utf8"
 )
 
 // Clean escapes the UTF8 encoded string provided as input so it is safe to print on a unix terminal.
-// It removes non printing characters and substitutes the vt100 escape character 0x1b with '^'.
+// It removes non printing characters and substitutes the vt100 escape character 0x1b with '^['.
 func Clean(s string) string {
-	return strings.Map(func(r rune) rune {
+	return replace(func(r rune) rune {
 		if r >= 32 && r != 127 { // Values below 32 (and 127) are special non printing characters (i.e. DEL, ESC, carriage return).
 			return r
 		} else if r == '\n' || r == '\t' { // Allow newlines and tabs.
 			return r
-		} else if r == 0x1b { // 0x1b denotes the start of a vt100 escape sequence. Substiture it with '^' (this is how it is usually shown, i.e. in vim).
-			return '^'
+		} else if r == 0x1b { // 0x1b denotes the start of a vt100 escape sequence. Substiture it with '^[' (this is how it is usually shown, i.e. in vim).
+			return -1
 		}
-		return -1
+		return -2
 	}, s)
 }
 
+// replace returns a copy of the string s with all its characters modified
+// according to the mapping function.
+// If mapping returns -1, the character is substituted with the two character string `^[`.
+// If mapping returns any other negative value, the character is dropped from the string with no replacement.
+// This function is similar to strings.Map, except for how -1 is handled.
+func replace(mapping func(rune) rune, s string) string {
```

Comment:
why can't you use strings.Map?

Reply:
strings.Map was the previous solution. The point is that strings.Map only allows to map one rune to one rune. I want to substitute `0x1b` (one rune) with `^[` (two runes), so I had to slightly modify strings.Map for this purpose. 

Another option would have been using strings.Map to escape everything else but `0x1b`, and then take care of `0x1b` -> `^[` with strings.Replace, but this would require two passes over the data instead of a single one.

Reply:
Makes me nervous to not just use strings.Map (which is battle tested by thousands of users) as opposed to a custom function.

Reply:
So do you think I should go with the second option and do two passes over the data?

For what is worth, I copied over all the tests which are given for strings.Map to check that this implementation passes them as well. The changes (w.r.t. strings.Map) are quite small. I don't have a strong preference here...

Reply:
Can `strings.Replace` just work with "0x1b" -> "^[", in one pass?

Reply:
OK, got it!

Reply:
OK, i think it's fine to give this a shot.  I worry that if we make two passes over the data, we're slowing down the terminal in a way that we can't easily detect, but might be perceivable.  This is basically `strings.Map`, and I realize we're forking it, but it probably changes very slowly at this point.

Reply:
In regards to two passes, aren't you doing two passes now?  line 39 and line 78.

Reply:
nvm, i see...

Reply:
I'm ok with merging, just worried that we will miss upstream bugfixes by forking the function out.

Can you add a comment that this is copied from strings.Map function and a comment on where it is different?  So we can maintain it easier in the future.

Reply:
Comments are added. I will merge after CI passes unless I hear any objections.

Thanks a lot for the review!